feat: Use Narwhals to allow PyArrow / Polars / Modin input

MarcoGorelli opened the pull request:
All pull requests must have an associated issue in the issue tracker. If there
isn't one, please go open an issue describing the defect, deficiency or desired
feature. You can read more about our issue and PR processes in the
[wiki](https://github.com/bokeh/bokeh/wiki/BEP-1:-Issues-and-PRs-management).

- [ ] issues: fixes #xxxx
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

MarcoGorelli committed:
feat: Use Narwhals to allow PyArrow / Polars input

MarcoGorelli committed:
revert package-lock, allow for tests to run without optional deps installed

MarcoGorelli committed:
minimal pandas version doesnt support `.convert_dtypes(dtype_backend="numpy_nullable")`

mattpap committed:
Upgrade to TypeScript 5.7 (#14167)

* Upgrade to TypeScript 5.7

* Update ndarray() to use ArrayBufferLike

FBruzzesi commented on tests/conftest.py:
```diff
@@ -48,3 +53,39 @@ def pytest_addoption(parser: _pytest.config.argparsing.Parser) -> None:
         "--no-js", action="store_true", default=False,
         help="only run python code and skip js")
+
+def pandas_constructor(obj) -> IntoDataFrame:
+    return pd.DataFrame(obj)  # type: ignore[no-any-return]
```
Thinking out loud here. I believe that we should eventually move these into narwhals (or something like narwhals-testing?). It's the 2nd/3rd time we add such constructors to a repo.

Maybe we could mark [narwhals#739](https://github.com/narwhals-dev/narwhals/issues/739) as medium priority

MarcoGorelli committed:
update docs

MarcoGorelli committed:
use Object("narwhals.DataFrame") instead of Object("pandas.Dataframe")

MarcoGorelli committed:
make more generic (support other pandas-like libraries)

MarcoGorelli committed:
deprecate PandasDataFrame

MarcoGorelli committed:
dont forget deprecated directive

MarcoGorelli committed:
mypy

mattpap committed:
Don't fail on partially updated Pane's view storage (#14170)

mattpap committed:
Fix spelling across the repository (#14176)

* Fix spelling across the repository

* Update baseline file names

* Add deprecation for misspelled bokehjs theme

henricryden committed:
correct help text for Block glyph coordinates (#14174)

* correct help text for Block glyph coordinates

* specify lower-left corner for block glyph

* fix syntax error in block example code

* Update src/bokeh/models/glyphs.py

Co-authored-by: Moritz Schreiber <68053396+mosc9575@users.noreply.github.com>

* Update src/bokeh/models/glyphs.py

Co-authored-by: Moritz Schreiber <68053396+mosc9575@users.noreply.github.com>

---------

Co-authored-by: Moritz Schreiber <68053396+mosc9575@users.noreply.github.com>

mattpap committed:
Add support for custom scatter markers (#14165)

* Add support for custom scatter markers

* Don't use webgl if custom markers are involved

* Improve Path2D related visuals' API

Commit:
Deployment updates for release 3.7.0.dev4

Commit:
Merge deployment staging branch staging-3.7.0.dev4

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/branch-3.7' into narwhals

maximlt committed:
add title to logo link to display the Bokeh version (#14185)

mattpap committed:
Cast numpy's number types to Python (#14187)

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/branch-3.7' into narwhals

mattpap committed:
Add support for stateful action tools (#14143)

* Allow ClickButtons to have state

* Display ExamineTool's dialog state

* Display FullscreenTool's state

* Allow CustomAction to maintain state

* Add legend_visibility_tool.py example

* Add fullscreen tool to color_scatter.py

* Allow to manage initial state

* Update image baselines

* Add release notes

* Improve CustomAction.active_callback docstring

* Improve gridplot()'s signature

* Improve types of CustomAction's callbacks

* Add legend visibility tool example to bokehjs

* Add unit tests

* Add a warning if active_callback didn't return a boolean

* Allow to set active state of CustomAction tools

* Add more unit tests

* Add alternatives to examples

* Simplify logic related to `callback` returning a boolean

* Correctly override internal property status

* Improve handling of null | undefined in gridplot()

* Fix syntax in docstrings

* Improve robustness of View.ready promise

* Robustify tool tap gesture in unit tests

MarcoGorelli committed:
do narwhals wrapping in columndata property

MarcoGorelli committed:
do narwhals wrapping in columndata property

MarcoGorelli committed:
restore test_instance

MarcoGorelli committed:
undeprecate PandasDataFrame, rename pd.py

MarcoGorelli committed:
tests

bryevdv committed:
Update README.md

mosc9575 committed:
fix unclear docs useing scatter() instead of circle() (#14194)

* use scatter() instead of cirlce()

* Update scatter.rst and Scatter in glyphs.py

* replace circle() with scatter() and fix typos

* some more replacement

mattpap committed:
Fix typing error and improve error message (#14206)

bryevdv committed:
Use \boldsymbol for bold math text (#14202)

* Use \boldsymbol for bold math text

* Update image baselines

---------

Co-authored-by: Mateusz Paprocki <mattpap@gmail.com>

orionlee committed:
app.rst - Fixed empty plot in the sample codes of section Updating fom unlocked callbacks (#14197)

* app.rst - Fixed empty plot in the sample codes of section Updating from unlocked callbacks

* app.rst - replace .circle() calls with .scatter()

* Update app.rst - simplify scatter() call per review

hoxbro committed:
Add `BOKEH_CHROME` environment variable for BokehJS test (#14209)

bryevdv committed:
Simple spatial index for wedge-like glyphs (#14212)

Commit:
Update README.md (#14216)

* Update README.md

* Updated CoC link to a generic link

* Update README.md

revert changes and update coc link to a generic link

mattpap committed:
Connect signals in DOM elements/nodes (#14132)

* Connect signals in DOM elements/nodes

* Add notes about API changes

Commit:
Deployment updates for release 3.7.0.dev5

Commit:
Merge deployment staging branch staging-3.7.0.dev5

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/branch-3.7' into narwhals

muendlein committed:
enhanced the default datetimeformatter with a dynamic one (#13854)

* enhanced the default datetimeformatter with a dynamic one

* adjusted defaults

* avoid InstanceDefault

* removed unnecessary imports

* fixed order

* Update src/bokeh/models/formatters.py

Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

* fixed formatting

* removed separate kwargs function

* renamed function

* added release note entry

---------

Co-authored-by: muendlein <max.muendlein@stud-mail.uni-wuerzburg.de>
Co-authored-by: Bryan Van de Ven <bryan@bokeh.org>

designerzim committed:
Update visjs link to working page of examples (#14227)

Fixing #14226 - Surface3D example code for wrapping.py contains an outdated link

gabalafou committed:
Add test to capture effect of setting max-height to 100% (#14219)

mattpap committed:
Improve typing in plotting._renderer (#14233)

mattpap committed:
Don't apply min_width/min_height unnecessarily (#14234)

mosc9575 committed:
replace stadiamaps with opentopomap in airports_map.py (#14237)

* replace stadiamaps with opentopomap in airports_map.py

mosc9575 committed:
14180 fix broken links (#14238)

* fixed some dead links

* remove stocks and add population server example

* fix link to palettes.py

* remove docs for python integration tests

* update link for dash-list

* update final links

* fix population ulr

* add pics as png

* replace mybinder by bokeh/tutorial

* replace more links to mybinder

* remove link to mybinder in jupyter.rst

---------

Co-authored-by: Dylan Cheng <chengdylan02@gmail.com>

gabalafou committed:
Add Google Chrome for Testing to list of Chrome executable names (#14239)

gabalafou committed:
Align canvas panel items to start, remove max-content legend height (#14191)

mattpap committed:
Fix link syntax in first_steps_7.rst (#14243)

gabalafou committed:
Build docs without GOOGLE_API_KEY if BOKEH_DOCS_CDN=local (#14240)

bryevdv committed:
Fix data_table sort performance (#14262)

* add argsort function to BokehJS

* use argsort on table index once for comparator

* construct explicit lookup to accommodate views

* Apply suggestions from code review

Co-authored-by: Mateusz Paprocki <mattpap@gmail.com>

---------

Co-authored-by: Mateusz Paprocki <mattpap@gmail.com>

mattpap committed:
Add support for manually generated stub files (*.pyi) (#13900)

* Add support for manually generated stub files (*.pyi)

* Properly declare visual properties

* Add release notes

* Add copyright header to *.pyi files

mattpap committed:
Convert CSS4 colors to CSS3 in SVG backend (#14264)

mattpap committed:
Allow to access tools via a context menu (#14228)

* Allow to access tools via a context menu

* Simplify menu item related model hierarchy

* Add release notes

* Update cross integration tests

* Ignore deprecation warnings in constructibility tests

* Allow None/null in place of DividerItem

* Migrate tools to using MenuItem model

* Improve disabled menu item cursor

* Add integration tests

* Prepare for key binding handling

* Extract preliminary CSS variable API

* Make the example more interactive

* Don't cancel native menu if custom menu didn't show

* Improve usage of range() iterator

* Add support for UIElement.context_menu == "auto"

* Rename {menus->legacy_menus}.less and menus{_->}.less

* Don't display icon column when not using icons

* Update type declarations

spenceforce committed:
Add return annotation to wrapped glyph/marker renderer methods (#14253)

* include return annotation in wrapped function signature

* add tests for meta data of wrapped glyph renderers

* remove redundant attribute setter operation

* fix ruff error of missing import

mattpap committed:
Assign labels instead of title in issue templates (#14250)

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/branch-3.7' into narwhals

MarcoGorelli committed:
use properties

MarcoGorelli committed:
move from_native into prepare_value, use properties instead of Object

MarcoGorelli committed:
test EagerSeries

MarcoGorelli committed:
rename test file

MarcoGorelli committed:
fix typing and test error message

MarcoGorelli committed:
set dtype in pd.Series constructor to avoid pandas futurewarning

MarcoGorelli committed:
bs4 compat: address camelcase deprecation warning (#14273)

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/branch-3.7' into narwhals

MarcoGorelli committed:
noop

MarcoGorelli committed:
move from_native into EagerDataFrame and EagerSeries prepare_value

MarcoGorelli committed:
it gets simpler
